### DISCUSSION TITLE
Support SSO login in the SDK

### DISCUSSION THREAD

**Discussion by br41nslug** [directus/directus#19723] [Feature Requests]
### Summary

Logging in with external providers requires opening an URL in the current or a new tab then be redirected to an external service and be directed back to the app authenticating.

Since the SDK uses AJAX requests this redirection flow is not currently supported.

### Basic Example

Something along the lines of

```ts
client.sso('google')
```

### Motivation

SSO is a good and broadly used alternative to user/password logins.

### Detailed Design

I am not sure about the detailed design or final flow. TBD

### Must Have:
- Support for various SSO techniques and vendors
- Docs on how to implement SSO with the SDK
- 

### Drawbacks

- There are likely different behaviors across the various idp vendors to account for

### Alternatives

Alternatively we could not support SSO like the current and old SDK

### Adoption Strategy

As new new feature without breaking change this would not require additional adoption for existing codebases.

### Unresolved Questions

- How to get the SSO to redirect to the frontend using the SDK instead of the Directus instance.
- How to read the incoming credentials once redirected back.

**Comment by cdwmhcc** [comment]
I think this is an interesting feature request.

**Comment by rijkvanzanten** [comment]
> Something along the lines of
> 
> `client.sso('google')`

I think we'll have to keep the difference between provider types in mind. LDAP logins will not be the same redirect flow as openid/oauth2. Maybe we should have a `client.openid('google')` type thing?

**Comment by alexey-yarmosh** [comment]
Isn't that a must have feature for every project that has separate FE and Directus as BE?
Looks like there is no way to implement SSO auth other than creating custom redirect endpoint (https://github.com/directus/directus/discussions/9347), which has its own drawbacks.

**Comment by NilsBaumgartner1994** [comment]
This is something we wish for around 1,5-2 years. Please

**Comment by alexey-yarmosh** [comment]
I was able to implement that by making directus sso to redirect back to FE app.
```js
const redirect = `${window.location.origin}${'/'}`;
await navigateTo(`http://${DIRECTUS_BE}/auth/login/${PROVIDER}?redirect=${encodeURIComponent(redirect)}`, { external: true });
```
That sets the refresh token to your cookies, so `directus.refresh()` can be called to bring access_token to the app. [source](https://github.com/directus/directus/discussions/9347#discussioncomment-1576512)

**Reply by NilsBaumgartner1994** [reply]
On native you cannot access the cookie of the Webbrowser?

**Reply by alexey-yarmosh** [reply]
You cannot access `directus_refresh_token` from js, which is good. But it will be auto included with the call to /auth/refresh so it will succeed.

**Reply by NilsBaumgartner1994** [reply]
Well how will it be auto included? The native app does not share the same cookies as the browser of the smartphone?

The login process is:
Native App: Open SSO login link
Smartphone: Leaves Native App and opens Browser
Browser: Gets authenticated and recieves cookie and redirects to native app
Native App: Was opened by DeepLink. No access to browser cookie --> No success on auth/refresh

Temporary workaround to get the refresh_token: https://github.com/directus/directus/discussions/22427

**Reply by alexey-yarmosh** [reply]
That solution is only for browsers, no info about native apps.

**Reply by NilsBaumgartner1994** [reply]
I know. But native apps would benefit if you could use such a login :D

**Comment by NilsBaumgartner1994** [comment]
Please also think about supporting SSO withing native apps.

**Comment by kopino4** [comment]
I am trying to find a solution for my Flutter mobile app. But it would be the same for native Android/Swift. Did you make it work by using the hacky solution you provided here? 

I already obtain an access token in my app using https://pub.dev/packages/google_sign_in

So i wonder how can i integrate it into Directus

https://github.com/directus/directus/discussions/22427

**Comment by NilsBaumgartner1994** [comment]
Yes. For me it currently works perfect

**Comment by JamesW1** [comment]
Thanks for the suggestion! We've moved feature requests over to our public roadmap at https://roadmap.directus.io/roadmap

I've migrated this request there for you: https://roadmap.directus.io/roadmap?id=a68db6cf-d9fa-4570-88d8-b9944454518a

Moving forwards, any updates on this subject will be visible on our public roadmap.
